### PR TITLE
Collect deopt info during guard failures.

### PR DESCRIPTION
**Pull request by ptersilie**
In the guard failure codegen, we use the live JIT variables stored inside the guard to ask the register allocator for their locations, and store this information in the `CompiledTrace`. When `__yk_deopt` is later called during trace execution we can retrieve this information to be used during deoptimisation.

### CODE REVIEW

File: ykrt/src/compile/jitc_yk/codegen/x86_64.rs [hunk 1]
```diff
@@ -49,7 +49,14 @@ const SYSV_CALL_STACK_ALIGN: usize = 16;
...
 
-extern "C" fn __yk_deopt(_ctrlpvars: *const c_void, _frameaddr: *const c_void) {
+extern "C" fn __yk_deopt(_ctrlpvars: *const c_void, _frameaddr: *const c_void, deoptid: usize) {
+    let ctr = crate::mt::THREAD_MTTHREAD
+        .with(|mtt| mtt.running_trace().unwrap())
+        .as_any()
+        .downcast::<X64CompiledTrace>()
+        .unwrap();
+    debug_assert!(deoptid < ctr.deoptinfo.len());
+    eprintln!("deopt frames: {:?}", ctr.deoptinfo[deoptid].frames);
```
Comment (ltratt) [inline ykrt/src/compile/jitc_yk/codegen/x86_64.rs:59]: Debug?

File: ykrt/src/compile/jitc_yk/codegen/x86_64.rs [hunk 2]
```diff
@@ -501,19 +511,45 @@ impl<'a> X64CodeGen<'a> {
         self.reg_into_new_local(inst_idx, WR0);
     }
 
+    fn codegen_arg(&mut self, inst_idx: InstrIdx, idx: u16) {
+        // For arguments passed into the trace function we simply inform the register allocator
+        // where they are stored and let the allocator take things from there.
+        match idx {
```
Comment (vext01) [inline ykrt/src/compile/jitc_yk/codegen/x86_64.rs:517]: Is this deciding which register to use for the ABI based on arg idx?

You can search `ARG_REGS` for this.

File: ykrt/src/compile/jitc_yk/codegen/x86_64.rs (hunk 1 above)
Comment (ptersilie) [reply ykrt/src/compile/jitc_yk/codegen/x86_64.rs:59]: This is in lieu of ways to test this properly. I want to check that the deoptinfo is actually there and has the right content. Maybe the `debug_assert` is enough here?

Comment (ltratt) [reply ykrt/src/compile/jitc_yk/codegen/x86_64.rs:59]: Yes, I think just the `debug_assert` is the right thing to do.

File: ykrt/src/compile/jitc_yk/codegen/x86_64.rs [hunk 3]
```diff
@@ -537,10 +573,22 @@ impl<'a> X64CodeGen<'a> {
...
+    frames: Vec<u64>,
+    // Vector of live JIT variable locations.
+    lives: Vec<LocalAlloc>,
+}
+
 #[derive(Debug)]
 pub(super) struct X64CompiledTrace {
     /// The executable code itself.
     buf: ExecutableBuffer,
+    /// Vector of boxed deopt info, tracked here so they can be freed when the compiled trace is
```
Comment (vext01) [inline ykrt/src/compile/jitc_yk/codegen/x86_64.rs]: Is "boxed" right? Don't see any `Box::new()` in this diff.

File: ykrt/src/compile/jitc_yk/codegen/x86_64.rs (hunk 1 above)
Comment (vext01) [reply ykrt/src/compile/jitc_yk/codegen/x86_64.rs:59]: I think he wants to match the output with fm until proper deopt is implemented?

File: ykrt/src/compile/jitc_yk/codegen/x86_64.rs (hunk 3 above)
Comment (vext01) [reply ykrt/src/compile/jitc_yk/codegen/x86_64.rs]: Ah, I see. You fixed this :)

File: ykrt/src/compile/jitc_yk/codegen/x86_64.rs (hunk 2 above)
Comment (ptersilie) [reply ykrt/src/compile/jitc_yk/codegen/x86_64.rs]: Try [f6a4955](https://github.com/ykjit/yk/pull/1052/commits/f6a4955905df99338035c19b004174c02fd70758)

File: ykrt/src/compile/jitc_yk/codegen/x86_64.rs (hunk 1 above)
Comment (ltratt) [reply ykrt/src/compile/jitc_yk/codegen/x86_64.rs:59]: Ah I see!

**Comment by ltratt** [comment]
Please squash.

**Comment by ptersilie** [comment]
> Please squash.

Just waiting for @vext01 to sign off on those new commits.

**Comment by vext01** [comment]
Go ahead.

**Comment by ltratt** [comment]
Please squash.

**Comment by ptersilie** [comment]
Squashed.

**Comment by ptersilie** [comment]
Fixed tests. But looks like I also need to rebase now.

**Comment by ltratt** [comment]
Please squash.

**Comment by ptersilie** [comment]
Squashed and rebased.

**Comment by ltratt** [comment]
Looks like a transitory error.

**Comment by ptersilie** [comment]
Ok to squash?

**Comment by ltratt** [comment]
Please squash.

**Comment by ptersilie** [comment]
Squashed.

**Pull request merged**